rd-net/RdFramework/Impl/Serializers.cs

```diff
@@ -348,24 +324,75 @@ bool TryGetTypeMapping(Type type1, out RdId rdId)
     }
 
     
-    private readonly HashSet<Type> myRegisteredToplevels = new HashSet<Type>();
     public void RegisterToplevelOnce(Type toplevelType, Action<ISerializers> registerDeclaredTypesSerializers)
     {
-      var r = new ToplevelRegistration(toplevelType, registerDeclaredTypesSerializers);
-#if !NET35
-      var task = myBackgroundRegistrar.SendOrExecuteInline(r);
-      Assertion.Assert(task.IsCompleted, "task.IsCompleted: {0}", task.Status);
-#else
-      RegisterToplevelInternal(r);
- #endif
+      myBackgroundRegistrar.Queue(toplevelType, registerDeclaredTypesSerializers);
+    }
+
+    struct ToplevelRegistration
+    {
+      internal Type Type { get; }
+      internal Action<ISerializers> Action { get; }
+
+      public ToplevelRegistration(Type type, Action<ISerializers> action)
+      {
+        Type = type;
+        Action = action;
+      }
     }
 
-    private void RegisterToplevelInternal(ToplevelRegistration r)
+    private class BackgroundRegistrar
```

Comment:
It's a very strange "fix". Maybe it worth to understand why actor doesn't work as expected (if it really happens) rather then reinventing the wheel and make yet another "background processer"?

Reply:
> worth to understand why actor doesn't work as expected 

I've done this twice before. This is third time and have no idea how to fix it this time (and frankly speaking, have no desire to). IMO Rd.Actor is overcomplicated and I spent too much time figuring out that happened each time. 

Issue was caused by thread pool starvation and SendAsync from main thread. It is only a guess, since we cannot reproduce customer's issue on our side:
```
[TestCase(true)]
[TestCase(false)]
public void TestActorStarvation(bool withMainThread)
{
  var def = new LifetimeDefinition();
  var actor = new Actor<int>("TestActor", def.Lifetime, x => { });  
  if (withMainThread)
  {
    new Thread(() =>
    {
      Thread.Sleep(100);
      actor.SendBlocking(1);
    }) {IsBackground = true}.Start();
  }  for (int i = 0; i < 300; i++)
  {
    Task.Run(() =>
    {
      for (int j = 0; j < 10000; j++)
        actor.WaitForEmpty();
    });
  }  Task.Run(() => { }).Wait();
}
```

Reply:
Well, 
1) There is a Threal Pool in resharper that is defended against starvation. I don't remember exact name (maybe UnlimitedThreadPool) but it must be in JetBrains.Threading so you can move it here and make scheduler upon it.
BTW. It's easy to check starvation on .NET 45 - after 40 threads allocated each new thread requires 1 sec to initiate.
2) I don't know what do you mean about "overcomplicated" but reintventing wheel when you don't know exact reason of problem is not a good programming practice.

Reply:
This is how ThreadManagementUnit works in ReSharper. But it also have issues - it is unable to detect thread starvation in some cases so I had to add a method that will tell it that the threads needs to be considered blocked.

@korifey in my understanding we do not need actors to reduce JIT time. We just need to process the serializers on background thread in some advance to thier actual use. And if we did not succeed just instantiate it sync when requested. This can be done in a controlled manner without scheduling the tasks on a thread pool hpoing for the best

Reply:
@korifey 
1. You are proposing to overcome the caveats of actor's implementation by API consumers. Rd is used in many places nowadays: dotCover, Unit Testing, dotTrace, ReSharper OOP. Fixing all of them and tracking new misuses is beyond my pay grade.
2. "overcomplicated" is my personal opinion, as I mentioned. Any constructive feedback is appreciated.

Reply:
@sergeyQx I got your point. As I mentioned before there is an UnlimitedThreadPool in ReSharper that handles starvation issue (you've even reviewed my code back there) and I propose to move it into Rd and use it everywhere in place of "handmade background processor". This pool try to use system thread pool and fall back to creating new threads only if starvation happens. 
UnlimitedThreadPool has attached TaskScheduler (per my memories, if not you can write one) that can be used by Actor to schedule activities sequentially on background.

@ulex I understand that you have to fit current business needs. Business continuity is a priority. On the other hand, one of  Rd's opensourcing reasons was to replace different historical R# implementation of one thing (like background processing) by canonical one (Actor).  Now what I see is a small step towards turning Rd back to "zoo" of reinvented wheels.
Actor relies havily on internal task scheduler so I propose just to change it to some scheduler without starvation. Or you know other issues in Actor?

Reply:
The starvation itself was caused by synchronous await in Actor's code (waitForEmpy), see test I attached above. That I propose is to avoid this starvation cause, rather than treating the symptoms by throwing new threads. 
Although it may be worth to introduce some reusable thing.

Reply:
So, if you propose to make some `waitAndPump` instead of `waitForEmpty`, it's a good idea that can be implemented directly in Actor. No need to make separate BackgroundRegistrar. It's so called "helper" approach when waiting thread actively stealing work from a queue. Fatal drawback here is that you invoke some potentially third-party code under stack frame of waiter thread. Usually we close our eyes on this drawback when mixing synchronous and asychronous code.

Few words attached test. I think that using `Task.Run()` is potentially flawed because of inevitable starvation on .NETFramework if something blocks inside. `WaitForEmpty()` here is just one example, but you will receive starvation every time on long blocking operation. Since it's hard to forbid blocking wait in general, I propose to use UnlimitedThreadPool everywhere in production to fix this problem once and for all.

Reply:
> it's a good idea that can be implemented directly in Actor

Could you briefly describe how you see it? Processor isn't thread safe and there is no api to steal a task from an actor, so I thought it would be impossible without tremendeous hacks

Meanwhile I've updated PR and introduce StealingScheduler to make code cleaner and reusable if necessary.

> Since it's hard to forbid blocking wait in general, I propose to use UnlimitedThreadPool 

Thanks for clarifying. blocking can be avoided in this particular scenario and I believe it is worth it. And I have no control over the code that uses `Task.Run`, the attached test is only a model based on client side snapshot.